File: Array/maxProductSubarray.py
import math

def maxProduct(nums):
    maxProd = nums[0]
    minProdEnd = nums[0]
    maxProdEnd = nums[0]
    
    for i in range(1, len(nums)):
        a = min(nums[i], minProdEnd*nums[i], maxProdEnd*nums[i])
        b = max(nums[i], minProdEnd*nums[i], maxProdEnd*nums[i])

        minProdEnd, maxProdEnd = a, b
        maxProd = max(maxProd, maxProdEnd)

    return maxProd

File: Array/maxProductSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxProduct = function(nums) {
    var max = nums[0];
    var minProdEnding = nums[0];
    var maxProdEnding = nums[0];
    
    for (let i=1; i<nums.length; i++) {
        var a = Math.min(nums[i], minProdEnding*nums[i], maxProdEnding*nums[i])
        var b = Math.max(nums[i], minProdEnding*nums[i], maxProdEnding*nums[i])
        minProdEnding = a
        maxProdEnding = b
        max = Math.max(max, maxProdEnding)
    }
    return max
};